### PR TITLE
Task 2 of the RS SCHOOL project - React, Unit testing

### PR DESCRIPTION
**Pull request by pokolga**
1. Task: https://github.com/rolling-scopes-school/tasks/blob/master/react/modules/tasks/tests.md
2. Screenshot: 
image

   
3. Deployment: -
4. Done 18.07.2025 / deadline 20.07.2025
5. Score: 100 / 100
- [x] Test Runner Setup (Jest/Vitest configured with RTL, coverage reporting works) - 15 points
- [x] Search Component Tests (localStorage, user input, search functionality) - 20 points
- [x] Results/CardList Component Tests (rendering, data display, loading states) - 15 points
- [x] Card Component Tests (rendering) - 15 points
- [x] Error Boundary Tests (error catching, fallback UI, test error button) - 15 points
- [x] API Integration Tests (mocked API calls, success/error handling) - 15 points
- [x] Husky Pre-push Hook (tests run automatically before push) - 5 points

### CODE REVIEW

File: src/__tests__/CardsList.test.tsx [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+  {
+    id: 2,
+    name: 'Morty Smith',
+    image: 'https://rickandmortyapi.com/morty.png',
+    species: 'Human',
+    status: 'Alive',
+  },
+];
+
+describe('CardsList', () => {
```
Comment (Margaryta-Maletz) [inline src/__tests__/CardsList.test.tsx:23]: It is better to add tests for additional scenarios:
- If the CardsList component has a loading state (e.g., when data is being fetched), it should be tested to ensure proper rendering during loading.
- A test should be added to verify the behavior when the results prop is empty or undefined. For example, does the component display a message like "no results"?
- If the component handles errors (e.g., failed API calls), a test should verify that the error state is displayed correctly.
- Test edge cases, such as very long character names, missing image URLs, or unexpected data formats.

You can implement all of these scenarios or some of them.

File: src/__tests__/Card.test.tsx [hunk 2]
```diff
@@ -0,0 +1,28 @@
+import { render, screen } from '@testing-library/react';
+import { Card } from '../components/card';
+import { describe, expect, it } from 'vitest';
+import type { Character } from '../models/types';
+
+describe('Card', () => {
```
Comment (Margaryta-Maletz) [inline src/__tests__/Card.test.tsx:6]: There is no test case to verify how the Card component behaves when some or all props are missing. For example:
- What happens if character.image or character.name is undefined?
- Does the component render fallback content or handle errors gracefully?
You can implement something like 
```
it('handles missing props gracefully', () => {
    const incompleteCharacter = { id: 2, name: '', image: '', species: '', status: '' };
    render(<Card character={incompleteCharacter} />);

    const img: HTMLImageElement = screen.getByRole('img');
    expect(img).toHaveAttribute('src', '');
    expect(img).toHaveAttribute('alt', '');

    const nameText: HTMLElement = screen.queryByText('');
    expect(nameText).not.toBeInTheDocument();

    const speciesStatusText: HTMLElement = screen.queryByText(/ — /);
    expect(speciesStatusText).not.toBeInTheDocument();
  });
```

File: src/__tests__/errorBoundary.test.tsx [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+      name: /Click to trigger error/i,
+    });
+    await userEvent.click(button);
+
+    const fallbackText: HTMLElement = screen.getByText('Something went wrong...');
+    const errorMessage: HTMLElement = screen.getByText('The test Error button was clicked!');
+    const tryAgainButton: HTMLButtonElement = screen.getByRole('button', {
+      name: /Try again!/i,
+    });
+
```
Comment (Margaryta-Maletz) [inline src/__tests__/errorBoundary.test.tsx:25]: The test mocks console.error to suppress error logs but does not verify that the error is logged correctly.
Improvement: Add an assertion to check that console.error is called with the expected error message.
```
vi.spyOn(console, 'error').mockImplementation(() => {});
expect(console.error).toHaveBeenCalledWith(expect.stringContaining('The test Error button was clicked!'));
```

**Review by Margaryta-Maletz — APPROVED** [review]
Great job!
But some comments on how to improve your tests you might want to consider and fix
Score: 100 / 100
  - [x] (15/15) Test Runner Setup (Jest/Vitest configured with RTL, coverage reporting works)
  - [x] (20/20) Search Component Tests (localStorage, user input, search functionality)
  - [x] (15/15) Results/CardList Component Tests (rendering, data display, loading states)    
  - [x] (15/15) Card Component Tests (rendering)
  - [x] (15/15) Error Boundary Tests (error catching, fallback UI, test error button)
  - [x] (15/15) API Integration Tests (mocked API calls, success/error handling)
  - [x] (5/5) Husky Pre-push Hook (tests run automatically before push)

File: src/__tests__/Card.test.tsx (hunk 2 above)
Comment (pokolga) [reply src/__tests__/Card.test.tsx:6]: I add the test like your one

File: src/__tests__/CardsList.test.tsx (hunk 1 above)
Comment (pokolga) [reply src/__tests__/CardsList.test.tsx:23]: I have a check in the logic of the Results component before displaying the cards that the array is not empty (and I cover it with the Results component test). Therefore, I think that CardsList in principle cannot receive an empty array

File: src/__tests__/errorBoundary.test.tsx (hunk 3 above)
Comment (pokolga) [reply src/__tests__/errorBoundary.test.tsx:25]: I have did it
